File: 01-programming/go/ejercicios/proyectos/proyecto-final/starter/modelos/modelos.go
// Package modelos define las entidades de dominio de la biblioteca.
package modelos

// GeneroLibro es un enum de los géneros disponibles.
type GeneroLibro string

const (
	Ficcion    GeneroLibro = "ficcion"
	NoFiccion  GeneroLibro = "no_ficcion"
	Ciencia    GeneroLibro = "ciencia"
	Tecnologia GeneroLibro = "tecnologia"
	Historia   GeneroLibro = "historia"
	Otro       GeneroLibro = "otro"
)

// Libro representa un libro de la biblioteca.
type Libro struct {
	ID         int
	Titulo     string
	Autor      string
	ISBN       string
	Anio       int
	Genero     GeneroLibro
	Disponible bool
}

// SetID permite al repositorio asignar el ID automáticamente.
func (l *Libro) SetID(id int) { l.ID = id }

// Miembro representa a una persona registrada en la biblioteca.
type Miembro struct {
	ID       int
	Nombre   string
	Email    string
	Telefono string
	Activo   bool
}

// SetID permite al repositorio asignar el ID automáticamente.
func (m *Miembro) SetID(id int) { m.ID = id }

// Prestamo representa el préstamo de un libro a un miembro.
// FechaInicio y FechaDevolucion se guardan como cadenas ISO ("2006-01-02").
type Prestamo struct {
	ID              int
	IDLibro         int
	IDMiembro       int
	FechaInicio     string
	FechaDevolucion string
	Devuelto        bool
}

// SetID permite al repositorio asignar el ID automáticamente.
func (p *Prestamo) SetID(id int) { p.ID = id }

File: 01-programming/rust/ejercicios/proyectos/proyecto-final/starter/src/modelos.rs
//! Modelos define las entidades de dominio de la biblioteca.

/// Géneros disponibles para clasificar un libro.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GeneroLibro {
    Ficcion,
    NoFiccion,
    Ciencia,
    Tecnologia,
    Historia,
    Otro,
}

/// Libro representa un libro de la biblioteca.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Libro {
    pub id: u32,
    pub titulo: String,
    pub autor: String,
    pub isbn: String,
    pub anio: u32,
    pub genero: GeneroLibro,
    pub disponible: bool,
}

impl Libro {
    pub fn set_id(&mut self, id: u32) {
        self.id = id;
    }

    pub fn id(&self) -> u32 {
        self.id
    }
}

/// Miembro representa a una persona registrada en la biblioteca.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Miembro {
    pub id: u32,
    pub nombre: String,
    pub email: String,
    pub telefono: String,
    pub activo: bool,
}

impl Miembro {
    pub fn set_id(&mut self, id: u32) {
        self.id = id;
    }

    pub fn id(&self) -> u32 {
        self.id
    }
}

/// Préstamo representa el préstamo de un libro a un miembro.
/// `fecha_inicio` y `fecha_devolucion` se guardan como cadenas ISO ("2006-01-02").
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Prestamo {
    pub id: u32,
    pub id_libro: u32,
    pub id_miembro: u32,
    pub fecha_inicio: String,
    pub fecha_devolucion: String,
    pub devuelto: bool,
}

impl Prestamo {
    pub fn set_id(&mut self, id: u32) {
        self.id = id;
    }

    pub fn id(&self) -> u32 {
        self.id
    }
}